include/message_filters/sync_policies/latest_time.h

```diff
@@ -0,0 +1,283 @@
+  typedef std::deque<M6Event> M6Deque;
+  typedef std::deque<M7Event> M7Deque;
+  typedef std::deque<M8Event> M8Deque;
+  typedef std::vector<M0Event> M0Vector;
+  typedef std::vector<M1Event> M1Vector;
+  typedef std::vector<M2Event> M2Vector;
+  typedef std::vector<M3Event> M3Vector;
+  typedef std::vector<M4Event> M4Vector;
+  typedef std::vector<M5Event> M5Vector;
+  typedef std::vector<M6Event> M6Vector;
+  typedef std::vector<M7Event> M7Vector;
+  typedef std::vector<M8Event> M8Vector;
+  typedef Events Tuple;
+  typedef std::tuple<M0Deque, M1Deque, M2Deque, M3Deque, M4Deque,
+                     M5Deque, M6Deque, M7Deque, M8Deque> DequeTuple;
+  typedef std::tuple<M0Vector, M1Vector, M2Vector, M3Vector, M4Vector,
+                     M5Vector, M6Vector, M7Vector, M8Vector> VectorTuple;
+
+  LatestTime(uint32_t)
+  : parent_(0),
+    ros_clock_(RCL_ROS_TIME)
+  {
+  }
+
+  LatestTime(const LatestTime& e)
+  {
+    *this = e;
+  }
+
+  LatestTime& operator=(const LatestTime& rhs)
+  {
+    parent_ = rhs.parent_;
+    events_ = rhs.events_;
+    rates_ = rhs.rates_;
+    ros_clock_ = rhs.ros_clock_;
+
+    return *this;
+  }
+
+  void initParent(Sync* parent)
+  {
+    parent_ = parent;
+  }
+
+  template<int i>
+  void add(const typename std::tuple_element<i, Events>::type& evt)
+  {
+    RCUTILS_ASSERT(parent_);
+    
+    std::lock_guard<std::mutex> lock(data_mutex_);
+    
+    if(!received_msg<i>())
+    {
+      rates_.push_back(Rate(1U, ros_clock_.now()));
+      // wait until we get each message once to publish
+      // then wait until we got each message twice to compute rates
+      // NOTE: this will drop a few messages of the faster topics until
+      //       we get one of the slowest so we can sync
+      std::get<i>(events_) = evt;  // adding here ensures we see even the slowest
+                                   // message twice before computing rate
+      return;
+    }
+
+    std::get<i>(events_) = evt;
+
+    compute_hz(rates_[i]);
+    if(i == find_pivot() && is_full())
+    {
+      publish();
+    }
+  }
+
+private:
+  // assumed data_mutex_ is locked
+  void publish()
+  {
+    parent_->signal(std::get<0>(events_), std::get<1>(events_), std::get<2>(events_),
+                    std::get<3>(events_), std::get<4>(events_), std::get<5>(events_),
+                    std::get<6>(events_), std::get<7>(events_), std::get<8>(events_));
+  }
+
+  struct Rate
+  {
+    uint64_t count;
+    rclcpp::Time start;
+    double hz;
+    Rate(const uint64_t &count, const rclcpp::Time &start)
+      : count(count),
+        start(start),
+        hz(0.0)
+    {
+    }
+    bool operator>(const Rate &that) const
+    {
+      return this->hz > that.hz;
+    }
+  };
+
+  // assumed data_mutex_ is locked
+  void compute_hz(Rate &rate)
+  {
+    rate.count += 1;
+    rclcpp::Time now = ros_clock_.now();
+    double elapsed = (now-rate.start).seconds();
+    rate.hz = double(rate.count)/elapsed;
+
+    if(rate.count % 10000U == 0U)
```

Comment:
I'm trying to understand the motivation for resetting every 10k calls. Maybe it would be smoother to use a moving window and always look back at the Nth previous call?

Also, maybe `compute_hz` could be a member of `Rate`, but you'd have to pass in the current time...

Reply:
I changed to EMA on hz (removing `count`) with a heavy weight on new hz (0.9*new_hz + 0.1*old_hz) and it works, but I'm thinking this way I would I also need to detect a step change in message rates (in the event someone dynamically modifies pub rate) to reset the EMA.

Reply:
I added another test to capture behavior of the policy and the EMA when pivot message changes rate and new pivot is selected.

Also, I moved `compute_h` to the `Rate` class.